Source/Fluids/WarpXFluidContainer.cpp

```diff
@@ -0,0 +1,1854 @@
+
+    // Grab the grid spacing
+    WarpX &warpx = WarpX::GetInstance();
+    const Real dt = warpx.getdt(lev);
+    const amrex::Geometry &geom = warpx.Geom(lev);
+    const auto dx = geom.CellSizeArray();
+    const amrex::Real clight = PhysConst::c;
+#if defined(WARPX_DIM_3D)
+    amrex::Real cx = (dt/dx[0]);
+    amrex::Real cy = (dt/dx[1]);
+    amrex::Real cz = (dt/dx[2]);
+    amrex::Real cx_half = 0.5*(dt/dx[0]);
+    amrex::Real cy_half = 0.5*(dt/dx[1]);
+    amrex::Real cz_half = 0.5*(dt/dx[2]);
+#elif defined(WARPX_DIM_XZ)
+    amrex::Real cx_half = 0.5*(dt/dx[0]);
+    amrex::Real cz_half = 0.5*(dt/dx[1]);
+    amrex::Real cx = (dt/dx[0]);
+    amrex::Real cz = (dt/dx[1]);
+#elif defined(WARPX_DIM_RZ)
+    const auto problo = geom.ProbLoArray();
+    amrex::Real cx_half = 0.5*(dt/dx[0]);
+    amrex::Real cz_half = 0.5*(dt/dx[1]);
+    amrex::Box const& domain = geom.Domain();
+#else
+    amrex::Real cz = (dt/dx[0]);
+    amrex::Real cz_half = 0.5*(dt/dx[0]);
+#endif
+
+    amrex::BoxArray ba = N[lev]->boxArray();
+
+    // Temporary Half-step values
+#if defined(WARPX_DIM_3D)
+    amrex::MultiFab tmp_Q_minus_x( amrex::convert(ba, IntVect(0,1,1)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_x( amrex::convert(ba, IntVect(0,1,1)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_minus_y( amrex::convert(ba, IntVect(1,0,1)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_y( amrex::convert(ba, IntVect(1,0,1)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_minus_z( amrex::convert(ba, IntVect(1,1,0)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_z( amrex::convert(ba, IntVect(1,1,0)), N[lev]->DistributionMap(), 4, 1);
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+    amrex::MultiFab tmp_Q_minus_x( amrex::convert(ba, IntVect(0,1)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_x( amrex::convert(ba, IntVect(0,1)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_minus_z( amrex::convert(ba, IntVect(1,0)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_z( amrex::convert(ba, IntVect(1,0)), N[lev]->DistributionMap(), 4, 1);
+#else
+    amrex::MultiFab tmp_Q_minus_z( amrex::convert(ba, IntVect(0)), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_z( amrex::convert(ba, IntVect(0)), N[lev]->DistributionMap(), 4, 1);
+#endif
+
+    // Advection push
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for (MFIter mfi(*N[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+
+        // Grow the entire domain
+        amrex::Box box = mfi.validbox();
+        box.grow(1);
+
+        // Loop over a box with one extra gridpoint to avoid needing to communicate
+        // the temporary arrays
+        amrex::Box tile_box = mfi.growntilebox(1);
+
+        // Limit the grown box for RZ at r = 0, r_max
```

Comment:
Added a comment "        // Loop over a box with one extra gridpoint in the ghost region to avoid 
        // an extra MPI communication between the edge value computation loop and
        // the flux calculation loop"